cineast-core/src/main/java/org/vitrivr/cineast/core/features/OCRSearch.java

```diff
@@ -1,24 +1,510 @@
+   */
+  private Mat img2Mat(BufferedImage original) {
+    BufferedImage in = new BufferedImage(original.getWidth(), original.getHeight(), BufferedImage.TYPE_INT_RGB);
+    Graphics2D graphic = in.createGraphics();
+    try {
+      graphic.setComposite(AlphaComposite.Src);
+      graphic.drawImage(original, 0, 0, null);
+    }
+    finally {
+      graphic.dispose();
+    }
+
+    Mat out;
+    byte[] data;
+
+    out = new Mat(in.getHeight(), in.getWidth(), CvType.CV_8UC3);
+    data = new byte[in.getWidth() * in.getHeight() * (int) out.elemSize()];
+    int[] dataBuff = in.getRGB(0, 0, in.getWidth(), in.getHeight(), null, 0, in.getWidth());
+    for (int i = 0; i < dataBuff.length; i++) {
+      data[i * 3] = (byte) ((dataBuff[i] >> 0) & 0xFF);
+      data[i * 3 + 1] = (byte) ((dataBuff[i] >> 8) & 0xFF);
+      data[i * 3 + 2] = (byte) ((dataBuff[i] >> 16) & 0xFF);
+    }
+
+    out.put(0, 0, data);
+    return out;
+  }
+
+  private GrayU8 bufferedImage2GrayU8(SegmentContainer shot, int frame_index) {
+    return ConvertBufferedImage.convertFromSingle(
+            shot.getVideoFrames().get(frame_index).getImage().getBufferedImage(), null, GrayU8.class);
+  }
+
+  /**
+   * @param streamLast
+   * @param streamFirst
+   * @return The similarity of the streams by taking into account the spatial distance, the overlap, the text similarity, and the frame offset
+   */
+  private double getSimilarity(TextStream streamLast, TextStream streamFirst) {
+    Quadrilateral_F64 lastBox = streamLast.findLastBox();
+    Quadrilateral_F64 firstBox = streamFirst.findFirstBox();
+
+    double spatial_x = (Math.abs(lastBox.getA().x - firstBox.getA().x) + Math.abs(lastBox.getB().x - firstBox.getB().x) +
+            Math.abs(lastBox.getC().x - firstBox.getC().x) + Math.abs(lastBox.getD().x - firstBox.getD().x)) / 4;
+    double spatial_y = (Math.abs(lastBox.getA().y - firstBox.getA().y) + Math.abs(lastBox.getB().y - firstBox.getB().y) +
+            Math.abs(lastBox.getC().y - firstBox.getC().y) + Math.abs(lastBox.getD().y - firstBox.getD().y)) / 4;
+
+    double spatial_dis = (spatial_x / Math.max(Math.abs(lastBox.getB().x -lastBox.getA().x), Math.abs(firstBox.getB().x -firstBox.getA().x)))
+            +(spatial_y / Math.max(Math.abs(lastBox.getD().y - lastBox.getA().y), Math.abs(firstBox.getD().y - firstBox.getA().y)));
+
+    double inv_IoU = 1 - getIntersectionOverUnion(lastBox, firstBox);
+
+    double edit_dis = 1 - new JaroWinklerSimilarity().apply(streamLast.getText(), streamFirst.getText());
+
+    return spatial_dis + inv_IoU + 10*edit_dis + 0.2*Math.abs(streamFirst.getFirst() - streamLast.getLast());
+  }
+
+  /**
+   * transformString transforms the string by replacing certain characters with others
+   * @param str The string to be transformed
+   * @return The transformed string
+   */
+  private String transformString(String str) { return str.replace("i", "l").replace("l", "1").replace("a", "o");}
```

Comment:
Why would you replace those characters? Are there known bugs in the method?

Reply:
The way I understand this is that this is a simple way to introduce an invariance against some somewhat commonly confused characters.

Reply:
That makes total sense, in this case i guess some documentation would be helpful @alexandertheus

Reply:
Will do! Upper case "i" and lowercase "l", and 1 are almost visually indistinguishable and the text recognizer hence has a hard time distinguishing them. Same can be said about "a" and "o", although to a lesser extent.